src/runtime.rs

```diff
@@ -9,92 +13,100 @@ use crate::{
-                }),
+/// Initialize a new apartment threaded runtime.
+///
+/// This calls `CoInitializeEx` with `COINIT_APARTMENTTHREADED`
+///
+/// On success, it is up to the user to call [`uninitialize`] when the COM runtime is no longer need.
+/// On error, the user should not call [`uninitialize`].
+pub fn new_runtime(threading_model: ThreadingModel) -> Result<(), HRESULT> {
+    unsafe {
+        match CoInitializeEx(std::ptr::null_mut::<c_void>(), threading_model as u32) {
+            // S_OK indicates the runtime was initialized, S_FALSE means it was initialized
+            // previously. In both cases we need to invoke `CoUninitialize` later.
+            S_OK | S_FALSE => Ok(()),
 
-                // Any other result is considered an error here.
-                hr => Err(hr),
-            }
+            // Any other result is considered an error here.
+            hr => Err(hr),
         }
     }
+}
 
-    pub fn get_class_object(&self, iid: &IID) -> Result<ComRc<dyn IClassFactory>, HRESULT> {
-        let mut class_factory = std::ptr::null_mut::<c_void>();
-        let hr = unsafe {
-            CoGetClassObject(
-                iid as *const IID,
-                CLSCTX_INPROC_SERVER,
-                std::ptr::null_mut::<c_void>(),
-                &IID_ICLASS_FACTORY as *const IID,
-                &mut class_factory as *mut *mut c_void,
-            )
-        };
-        if FAILED(hr) {
-            return Err(hr);
-        }
-
-        Ok(unsafe { ComRc::from_raw(class_factory as *mut *mut _) })
+/// Get the class object with the associated [`IID`]
+///
+/// Calls `CoGetClassObject` internally
+pub fn get_class_object(iid: &IID) -> Result<ComRc<dyn IClassFactory>, HRESULT> {
+    let mut class_factory = std::ptr::null_mut::<c_void>();
+    let hr = unsafe {
+        CoGetClassObject(
+            iid as *const IID,
+            CLSCTX_INPROC_SERVER,
+            std::ptr::null_mut::<c_void>(),
+            &IID_ICLASS_FACTORY as *const IID,
+            &mut class_factory as *mut *mut c_void,
+        )
+    };
+    if FAILED(hr) {
+        return Err(hr);
     }
 
-    pub fn create_instance<T: ComInterface + ?Sized>(
-        &self,
-        clsid: &IID,
-    ) -> Result<ComRc<T>, HRESULT> {
-        unsafe {
-            Ok(ComRc::new(
-                self.create_raw_instance::<T>(clsid, std::ptr::null_mut())?,
-            ))
-        }
-    }
+    Ok(unsafe { ComRc::from_raw(class_factory as *mut *mut _) })
+}
 
-    pub fn create_aggregated_instance<T: ComInterface + ?Sized, U: CoClass>(
-        &self,
-        clsid: &IID,
-        outer: &mut U,
-    ) -> Result<ComPtr<T>, HRESULT> {
-        unsafe { self.create_raw_instance::<T>(clsid, outer as *mut U as *mut c_void) }
+/// Create an instance of a CoClass with the associated class id
+///
+/// Calls `CoCreateInstance` internally
+pub fn create_instance<T: ComInterface + ?Sized>(clsid: &IID) -> Result<ComRc<T>, HRESULT> {
```

Comment:
Is this safe to call if COM hasn't been initialized?

Reply:
Good point. I'll see if I can find this out